File: 08-Testing-Java/Solution/test/testing/tests/FakeLogger.java
package testing.tests;

import testing.ILogger;

import java.util.ArrayList;

public final class FakeLogger implements ILogger {
    private final ArrayList<String> loggedMessages = new ArrayList<>();

    @Override
    public void log(String message) {
        loggedMessages.add(message);
    }

    public String[] getLoggedMessages() {
        return loggedMessages.toArray(String[]::new);
    }
}

File: 08-Testing-CSharp/Solution/Testing.Tests/FakeLogger.cs
namespace Testing.Tests;

public sealed class FakeLogger : ILogger
{
    private readonly List<string> _loggedMessages = new();
    
    public void Log(string message) => _loggedMessages.Add(message);

    public IReadOnlyList<string> LoggedMessages => _loggedMessages.AsReadOnly();
}